geonetwork/gn3-migration/schema-validation/fix-mcp-schema-validation.xsl

```diff
@@ -0,0 +1,362 @@
+    <xsl:copy>
+      <xsl:choose>
+        <xsl:when test="(./*[local-name()='beginTime']) and not(./*[local-name()='endTime'])">
+          <xsl:apply-templates select="@*" />
+          <gmd:extent>
+              <gml:TimePeriod gml:id="{generate-id(.)}">
+                <gml:beginPosition>
+                    <xsl:apply-templates select="./*[local-name()='beginTime']/gco:DateTime/text()"/>
+                </gml:beginPosition>
+                <gml:endPosition/>
+              </gml:TimePeriod>
+            </gmd:extent>
+        </xsl:when>
+        <xsl:otherwise>
+          <xsl:apply-templates select="@* | node()" />
+      </xsl:otherwise>
+      </xsl:choose>
+    </xsl:copy>
+  </xsl:template>
+
+   <!--  4402cb50-e20a-44ee-93e6-4728259250d2  -->
+  <!--   Missing `<gmd:code>` element added  -->
+  <xsl:template match="gmd:RS_Identifier">
+    <xsl:copy>
+      <xsl:choose>
+        <xsl:when test="not(./gmd:code) and ((./gmd:codeSpace/gco:CharacterString='WGS84') or (./gmd:codeSpace/gco:CharacterString='GDA94'))">
+          <gmd:code>
+            <gco:CharacterString>
+              <xsl:apply-templates select="./gmd:codeSpace/gco:CharacterString/text()" />
+            </gco:CharacterString>
+          </gmd:code>
+        </xsl:when>
+        <xsl:otherwise>
+          <xsl:apply-templates select="@* | node()" />
+        </xsl:otherwise>
+      </xsl:choose>
+    </xsl:copy>
+  </xsl:template>
+
+  <!-- Missing `<gmd:date>` element added  -->
+  <!--  <xsl:template match="gmd:date/gmd:CI_Date[not(./gmd:date) and (./gmd:dateType/gmd:CI_DateTypeCode/@codeListValue='']" />-->
+  <xsl:template match="gmd:CI_Date">
+    <xsl:copy>
+      <xsl:if test="not(./gmd:date)">
+        <gmd:date gco:nilReason="missing"/>
+      </xsl:if>
+       <xsl:apply-templates select="@* | node()" />
+    </xsl:copy>
+  </xsl:template>
+
+  <!--  f47a6929-1f19-4724-b74b-7c8579872cb7-->
+  <!-- Removed unnecessary `<gml:duration>` element   -->
+  <xsl:template match="gmd:temporalElement/gmd:EX_TemporalExtent/gmd:extent/gml:TimePeriod/gml:duration[not(*) and not(normalize-space())]" />
+
+  <!--  bae2b8ce-e1b4-40ea-b7bb-4e4c8341b277-->
+  <!-- Removed empty `<gmd:function/gmd:CI_OnLineFunctionCode>` element   -->
+  <xsl:template match="gmd:CI_OnlineResource/gmd:function[./gmd:CI_OnLineFunctionCode/@codeList = '' and ./gmd:CI_OnLineFunctionCode/@codeListValue='']" />
+
+  <!-- Missing `<mcp:term>` element added  -->
+  <xsl:template match="mcp:parameterUnits/mcp:DP_Term" xmlns:mcp="http://schemas.aodn.org.au/mcp-2.0">
+    <xsl:copy>
+      <xsl:if test="not(./mcp:term)">
+        <mcp:term gco:nilReason="missing"/>
+      </xsl:if>
+       <xsl:apply-templates select="@* | node()" />
+    </xsl:copy>
+  </xsl:template>
+
+  <!--  064bff02-7ba3-46e0-b873-afdf387a1205  -->
+  <!-- Removed empty `<gmd:contact/gmd:CI_ResponsibleParty>` element   -->
+  <xsl:template match="gmd:contact[./gmd:CI_ResponsibleParty[not(*) and not(normalize-space())]]" />
```

Comment:
This line seems to be missing empty elements that still have attributes but no values. I'm not sure if we want to remove elements such as this as they are technically valid but have no value. There are 2 other tags where a specific handler for empty values was added to the 19115-3 transform (see https://github.com/aodn/backlog/issues/1604). These empty elements translate to invalid records in 19115-3 via Simon's original MCP transform. An example of a valid empty element is below:
```xml
              <gmd:citedResponsibleParty>
                <gmd:CI_ResponsibleParty>
                  <gmd:individualName gco:nilReason="missing">
                    <gco:CharacterString />
                  </gmd:individualName>
                  <gmd:organisationName gco:nilReason="missing">
                    <gco:CharacterString />
                  </gmd:organisationName>
                  <gmd:positionName gco:nilReason="missing">
                    <gco:CharacterString />
                  </gmd:positionName>
                  <gmd:contactInfo>
                    <gmd:CI_Contact>
                      <gmd:phone>
                        <gmd:CI_Telephone>
                          <gmd:voice gco:nilReason="missing">
                            <gco:CharacterString />
                          </gmd:voice>
                          <gmd:facsimile gco:nilReason="missing">
                            <gco:CharacterString />
                          </gmd:facsimile>
                        </gmd:CI_Telephone>
                      </gmd:phone>
                      <gmd:address>
                        <gmd:CI_Address>
                          <gmd:deliveryPoint gco:nilReason="missing">
                            <gco:CharacterString />
                          </gmd:deliveryPoint>
                          <gmd:city gco:nilReason="missing">
                            <gco:CharacterString />
                          </gmd:city>
                          <gmd:administrativeArea gco:nilReason="missing">
                            <gco:CharacterString />
                          </gmd:administrativeArea>
                          <gmd:postalCode gco:nilReason="missing">
                            <gco:CharacterString />
                          </gmd:postalCode>
                          <gmd:country gco:nilReason="missing">
                            <gco:CharacterString />
                          </gmd:country>
                          <gmd:electronicMailAddress gco:nilReason="missing">
                            <gco:CharacterString />
                          </gmd:electronicMailAddress>
                        </gmd:CI_Address>
                      </gmd:address>
                    </gmd:CI_Contact>
                  </gmd:contactInfo>
                  <gmd:role>
                    <gmd:CI_RoleCode codeList="http://bluenet3.antcrc.utas.edu.au/mcp-1.4/schema/resources/Codelist/gmxCodelists.xml#CI_RoleCode" codeListValue="" />
                  </gmd:role>
                </gmd:CI_ResponsibleParty>
              </gmd:citedResponsibleParty>
```

Reply:
@jonescc Do you suggest to add transform to fix empty attributes values even though the record is valid with empty attributes (`<gmd:CI_RoleCode codeList="http://bluenet3.antcrc.utas.edu.au/mcp-1.4/schema/resources/Codelist/gmxCodelists.xml#CI_RoleCode" codeListValue="" />`) ?

Reply:
No if its valid I wouldn't bother trying ot fix it.  Might be worth reporting for Nat to do a cleanup or for a bulk cleanup at some point.  Will be blowing out the scope of this task unnecessarily otherwise IMO.

Reply:
I think the transform should work with valid elements/attributes.   E.g. keeping nilReason codes etc.

Reply:
Unless that contradicts something I said earlier and which case let me know so I can apologise or consider more closely.

Reply:
Reported to Nat for review.